keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,842 @@
+   uint32 container_id = 1;
+   // host_id is the starting ID for the mapping on the host.
+   uint32 host_id = 2;
+   // size is the number of elements in the mapping.
+   uint32 size = 3;
+}
+
+// LinuxIDMappings represents the user and group user namespace ID mappings.
+message LinuxIDMappings {
+   // uid_mappings is an array of user ID mappings.
+   repeated LinuxIDMapping uid_mappings = 1;
+   // gid_mappings is an array of group ID mappings.
+   repeated LinuxIDMapping gid_mappings = 2;
+}
+
+// NamespaceOption provides options for Linux namespaces.
+message NamespaceOption {
+    ...
+    // User namespace for this container/sandbox.
+    // Note: There is currently no way to set CONTAINER scoped user namespace in the Kubernetes API.
+    // Namespaces currently set by the kubelet: POD, NODE
+    Namespace user = 5;
+    // IDs mappings to use if the user NamespaceMode is POD
+    LinuxIDMappings id_mappings = 6
+}
+```
+
+### Add userNamespaceMode Field
+
+The `userNamespaceMode` field can be added in two different places. This
+proposal presents the two possibilities to discuss with the community.
+
+<<[UNRESOLVED where to put the userNamespaceMode field ]>>
+
+#### Option 1: PodSpec
+
+Add it to `v1.PodSpec` following the rationale that other fields (`host{Network,
+IPC, PID}`) that control namespaces behaviour are defined in this place.
+
+```
+const (
+	UserNamespaceModeHost    PodUserNamespaceMode = "Host"
+	UserNamespaceModeCluster PodUserNamespaceMode = "Cluster"
+)
+
+type PodSpec struct {
+...
+  // UserNamespaceMode controls how user namespaces are used for this Pod.
+  // Three modes are supported:
+  // "Host": The pod shares the host user namespace. (default value).
+  // "Cluster": The pod uses a cluster-wide configured ID mappings.
+  // +k8s:conversion-gen=false
+  // +optional
+  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,36,opt  name=userNamespaceMode"`
+...
+```
+
+#### Option 2: PodSecurityContext
+
+Add it to `v1.PodSpec.PodSecurityContext` as this field controls a security
+aspect of the pod.
+
+```
+const (
+	UserNamespaceModeHost    PodUserNamespaceMode = "Host"
+	UserNamespaceModeCluster PodUserNamespaceMode = "Cluster"
+)
+
+type PodSecurityContext struct {
+...
+  // UserNamespaceMode controls how user namespaces are used for this Pod.
+  // Three modes are supported:
+  // "Host": The pod shares the host user namespace. (default value).
+  // "Cluster": The pod uses a cluster-wide configured ID mappings.
+  // +optional
+  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,11,opt  name=userNamespaceMode"`
+...
+```
+<<[/UNRESOLVED]>>
+
+### Configuring the Cluster ID Mappings
+
+This proposal considers two different ways to configure the ID mappings used for
+the `Cluster` mode. This is for discussion with the community and only one will
+be considered.
+
+<<[UNRESOLVED where to configure the cluster wide ID mappings ]>>
```

Comment:
We'll need to think through how this can change over time.  I know you want to say "don't change it" but experience says it will eventually need to be changed.,